### PR TITLE
Implement cask renames.

### PR DESCRIPTION
**Pull request by reitermarkus**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This adds support for a `cask_renames.json` file in taps.

To simplify the code, this also removes (most) support for multiple installed versions, which normally isn't possible anyways except when using `install --force` instead of `upgrade`. This was only used to give a warning when not all versions were uninstalled.

### CODE REVIEW

**Comment by reitermarkus** [comment]
@Rylan12, are formula renames supported with the API, if so, how does this work?

**Comment by apainintheneck** [comment]
Formula renames aren't supported by the API yet unfortunately.
- https://github.com/Homebrew/brew/issues/14897

**Review by MikeMcQuaid — COMMENTED** [review]
Nice work so far, very happy to see this! 🎉 

> update-report probably needs some changes still.

I think that'll need to be part of this PR (or at least implemented before we roll out any renames in homebrew-cask).

Similarly, I think `brew migrate` should be used/implemented so it behaves somewhat similarly to how it does for formulae.

Also, if you get the chance to fix how this is handled for formulae in the API too, we'd be very grateful!

Thanks again @reitermarkus! ❤️

**Comment by reitermarkus** [comment]
@Homebrew/cask would be great if someone could test this.

**Comment by bevanjkay** [comment]
@reitermarkus I have given this a bit of testing and all looks good to me. I used the `ankerslicer->ankermake` PR with adding the `cask_renames.json` and it seemed to reference the cask correctly. 

One question, should this work recursively? Theoretical example;

```json
{
    "ankerslicer": "ankermake",
    "anker": "ankerslicer"
}
```

`brew info anker` tries to resolve to `ankerslicer`.

File: Library/Homebrew/cmd/update-report.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/update-report.rb:493]: Are these equivalent?

File: Library/Homebrew/cmd/update-report.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/update-report.rb:781]: Given https://github.com/Homebrew/brew/blob/c6e438b3912b95ef5d75de9626bd4d065f071999/Library/Homebrew/cmd/update-report.rb#L679-L680 this may be worth skipping and we can make a minor release sooner rather than later?

File: Library/Homebrew/tap.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/tap.rb:702]: This won't work with the API so may be worth skipping?

**Review by MikeMcQuaid — COMMENTED** [review]
Looking good so far! Missing things I think are
- install from API support (a big one)
- `brew migrate` doing something sensible here

File: Library/Homebrew/cmd/update-report.rb
Comment (reitermarkus) [reply Library/Homebrew/cmd/update-report.rb:493]: It's a `Set`, so yes.

File: Library/Homebrew/tap.rb
Comment (reitermarkus) [reply Library/Homebrew/tap.rb:702]: Skip based on what?

File: Library/Homebrew/cmd/update-report.rb
Comment (reitermarkus) [reply Library/Homebrew/cmd/update-report.rb:781]: I don't quite get why it shouldn't possible with the API, as long as you have a previous API JSON to compare against?

File: Library/Homebrew/tap.rb
Comment (MikeMcQuaid) [reply Library/Homebrew/tap.rb:702]: I mean: I'm not sure it's worth implementing this logic for only on disk taps when we're primarily going to be using it for homebrew-cask to begin with which won't work like this because there's no `path` present on disk.

File: Library/Homebrew/cmd/update-report.rb
Comment (reitermarkus) [reply Library/Homebrew/cmd/update-report.rb:781]: I think we have the checksum of the formula/cask files.

File: Library/Homebrew/tap.rb
Comment (reitermarkus) [reply Library/Homebrew/tap.rb:702]: `fomula_renames` does the same thing which will also need to be changed in the same way to support the API. Also, I can only test this with a local tap as-is.

Comment (MikeMcQuaid) [reply Library/Homebrew/tap.rb:702]: Yeh, it needs implemented for formulae too. My point is: we need to implement this in the cask API before we can roll out any renames to users so: totally fine to test this as-is for now but will need API support before it can be used in any taps.

Comment (reitermarkus) [reply Library/Homebrew/tap.rb:702]: > before it can be used in any taps

I mean it can be used like this for third-party taps which aren't using the API (homebrew/cask-versions is not using the API, so we could actually already use it there and migrate the casks with the new name). Just because we will mainly be using it in homebrew/cask doesn't mean third-party taps shouldn't have support for this.

Even with API support, `Tap#cask_renames` will still need to exist but get the rename information from the API JSON instead I assume.

**Comment by reitermarkus** [comment]
> One question, should this work recursively?

How does it work for formulae?

File: Library/Homebrew/cmd/update-report.rb
Comment (MikeMcQuaid) [reply Library/Homebrew/cmd/update-report.rb:781]: If we already have it: I wouldn't be opposed to it being implemented and undeprecated.

File: Library/Homebrew/tap.rb
Comment (MikeMcQuaid) [reply Library/Homebrew/tap.rb:702]: > Just because we will mainly be using it in homebrew/cask doesn't mean third-party taps shouldn't have support for this.
> Even with API support, Tap#cask_renames will still need to exist but get the rename information from the API JSON instead I assume.

agreed 👍🏻

File: Library/Homebrew/cask/cask.rb
Comment (apainintheneck) [inline Library/Homebrew/cask/cask.rb:191]: Should `#installed_caskfile` and `#installed_version` get memoized here?

File: Library/Homebrew/test/cask/cask_spec.rb
Comment (apainintheneck) [inline Library/Homebrew/test/cask/cask_spec.rb:18]: Would it be possible to scope this down a little bit by maybe specifying the number of times we expect it to be called?

**Review by apainintheneck — COMMENTED** [review]
It'd be nice to see how the report output should look after these changes as well.

Regarding the question about whether the formula renames are recursive, I looked into it and I don't think they are. It looks to me like we just take the first old name for the current formula.

https://github.com/Homebrew/brew/blob/4fe3436682365f81eb68f56a2b8b5d766a372e00/Library/Homebrew/formula.rb#L495-L501

File: Library/Homebrew/cask/cask.rb
Comment (reitermarkus) [reply Library/Homebrew/cask/cask.rb:191]: No. There is actually a test which was silently broken because the installed version was cached.

Comment (apainintheneck) [reply Library/Homebrew/cask/cask.rb:191]: Not sure I understand. The version wasn't cached before, right? Which test was silently broken?

Comment (reitermarkus) [reply Library/Homebrew/cask/cask.rb:191]: I was thinking of a different issue. The issue was that previously `installed?` counted timestamped directories which don't contain a `.rb` or `.json`. That was a issue caused somewhere in `Installer`.

I still don't think it's a good idea to cache installed versions, since it'll probably break “install” and “uninstall” tests.

**Comment by apainintheneck** [comment]
This doesn't need to happen in this PR but the docs also need to be updated after this change eventually.
- https://docs.brew.sh/Rename-A-Formula

File: Library/Homebrew/test/cask/cask_spec.rb
Comment (reitermarkus) [reply Library/Homebrew/test/cask/cask_spec.rb:18]: Does that work with `allow_*` methods?

Comment (apainintheneck) [reply Library/Homebrew/test/cask/cask_spec.rb:18]: Doesn't look like it. I also see that this seems to be a common pattern in our test suite so ignore what I said.

It might be possible with `expect_any_instance_of` but that doesn't seem to be used in our codebase at all.

**Review by miccal — APPROVED** [review]

**Comment by reitermarkus** [comment]
I think we can merge this.

Adding API support should be done together with formulae.

Making `brew migrate` do something sensible can be done by someone who will use `brew migrate`.

**Review by MikeMcQuaid — CHANGES_REQUESTED** [review]
> Adding API support should be done together with formulae.
> 
> Making `brew migrate` do something sensible can be done by someone who will use `brew migrate`.

Sorry, I would like to see these done before this is merged or before I do a final pass of review (which I'd like to do before this is merged)

This feature is not useful without API support and making `brew migrate` behave appropriately (even if it just errors out saying that it shouldn't be run on casks) is also important.

**Comment by MikeMcQuaid** [comment]
> It's as useful as `formula_renames` is currently.

The important difference is that `formula_renames` was useful before API support.

---

If you'd rather this was merged first just to slim the diff: could at least the update-report logic be gated on `ENV["HOMEBREW_CASK_RENAMES"]` or similar for testing without end-users stumbling onto this before it works?

**Comment by reitermarkus** [comment]
I just realised I misunderstood what `brew migrate` does and I'm asking myself: Why does this command exist, doesn't migration happen in `brew upgrade`? 

In this PR, casks are migrated on `upgrade` (or `reinstall`), so `brew migrate` doesn't need to support casks at all, I guess?

**Comment by Bo98** [comment]
Migration doesn't necessarily involve a version update so the call in `brew upgrade` won't necessarily happen. With that said, `brew migrate` on it's own is indeed seldomly used as it's done automatically in `update-report`, and that's where most people will get the migration done.

**Comment by MikeMcQuaid** [comment]
> `brew migrate` is suggested by some errors to retry a failed migration (potentially with `--force`), which should hopefully never happen in practice often.
> 
> Similar story to `brew postinstall` really. You never need to invoke that directly most of the time, but it can be useful to retry a failed operation.

Agreed. I also think it's important that casks and formulae behave similarly here in terms of:
- when they are migrated
- how they are migrated
- what `brew migrate` does

To make them diverge (again) is going to just cause user confusion.

**Comment by MikeMcQuaid** [comment]
> > It's as useful as `formula_renames` is currently.
> 
> The important difference is that `formula_renames` was useful before API support.

Note: `formula_renames` now supports the API as-of https://github.com/Homebrew/brew/pull/15320 and, given discussion about `brew migrate` above: I'm back to thinking that this should have API, `brew migrate` support and a method like this for casks that's run at `brew update` time before it's merged:
https://github.com/Homebrew/brew/blob/1c29689e1342bbeed38d25e0eead1da4162f5f97/Library/Homebrew/cmd/update-report.rb#L589

File: Library/Homebrew/cask/cask.rb
Comment (apainintheneck) [inline Library/Homebrew/cask/cask.rb:86]:
```suggestion
        tap.cask_renames.select { |_, newtoken| newtoken == token }.keys
```

Nit: For clarity

This is just copied from the formula version.

File: Library/Homebrew/cask/cask.rb
Comment (apainintheneck) [inline Library/Homebrew/cask/cask.rb:152]: Should we check for the existence of the default installed caskroom path before entering this loop?

File: Library/Homebrew/cask/cask.rb
Comment (apainintheneck) [inline Library/Homebrew/cask/cask.rb:175]: Maybe a comment explaining this would be nice since it's not obvious at a glance.

File: Library/Homebrew/cask/cask.rb
Comment (apainintheneck) [inline Library/Homebrew/cask/cask.rb:1]: Nit: Should we change `#outdated_versions` to `#outdated_version` if we're only returning 0 or 1 versions?

File: Library/Homebrew/cask/migrator.rb
Comment (apainintheneck) [inline Library/Homebrew/cask/migrator.rb:48]: Are you sure we can't just do this here or is there some edge case I'm missing?

```suggestion
      new_installed_caskfile = old_installed_caskfile.sub(old_token, new_token)
```

**Review by apainintheneck — APPROVED** [review]
LGTM, I left some nits but feel free to ignore them if you disagree with them. The only other thing I wonder about is whether the warn flag used when loading casks or formulae should be tied somehow to the `--quiet` and `--verbose` options but I guess we can think about that some more if people start complaining about it.

**Comment by reitermarkus** [comment]
> the warn flag used when loading casks or formulae should be tied somehow to the `--quiet`

I think `--quiet` would make sense to be handled in `opoo` directly so that all warnings are hidden, regardless of where they are coming from.

File: Library/Homebrew/cask/migrator.rb
Comment (reitermarkus) [reply Library/Homebrew/cask/migrator.rb:48]: That's not the same thing since `sub` will only replace the first occurrence, which would be in `Caskroom/<token>/`, but we only want to replace it on the file name.

File: Library/Homebrew/cask/cask.rb
Comment (reitermarkus) [reply Library/Homebrew/cask/cask.rb:1]: Yes, but in another PR, since this will also require some changed in `brew outdated`.

File: Library/Homebrew/cask/cask.rb
Comment (reitermarkus) [reply Library/Homebrew/cask/cask.rb:152]: It shouldn't normally happen (probably only if migration fails for some reason), but old versions should be preferred in case a directory exists for both old and new names.

**Review by MikeMcQuaid — APPROVED** [review]
Approach makes sense to me, thanks for adding `brew update` and `brew migrate` support @reitermarkus and great work (again!) here. Will request a few more reviews just to make sure this looks good to folks who have touched this code recently.

Mild concern about some missing test coverage here but not blocking.

File: Library/Homebrew/cask/migrator.rb
Comment (apainintheneck) [reply Library/Homebrew/cask/migrator.rb:48]: They seem to print the same result on my machine when I tested them in the REPL.

```rb
~ $ brew irb
^[[A==> Interactive Homebrew Shell
Example commands available with: `brew irb --examples`
brew(main):001:0> cask = :virtualbox.c; nil
=> nil
brew(main):002:0> old_token = cask.token
=> "virtualbox"
brew(main):003:0> new_token = "virtualcats"
=> "virtualcats"
brew(main):004:0> old_installed_caskfile = cask.installed_caskfile.relative_path_from(cask.caskroom_path)
=> #<Pathname:.metadata/7.0.8,156879/20230423212244.154/Casks/virtualbox.rb>
brew(main):005:0> new_installed_caskfile = old_installed_caskfile.dirname/old_installed_caskfile.basename.sub(old_token, new_token)
=> #<Pathname:.metadata/7.0.8,156879/20230423212244.154/Casks/virtualcats.rb>
brew(main):006:0> new_installed_caskfile = old_installed_caskfile.sub(old_token, new_token) 
=> #<Pathname:.metadata/7.0.8,156879/20230423212244.154/Casks/virtualcats.rb>
```

**Comment by apainintheneck** [comment]
> > the warn flag used when loading casks or formulae should be tied somehow to the `--quiet`
> 
> I think `--quiet` would make sense to be handled in `opoo` directly so that all warnings are hidden, regardless of where they are coming from.

I really like this idea.

**Comment by reitermarkus** [comment]
We already do the same for `odebug` and `--debug`.

**Comment by MikeMcQuaid** [comment]
> I think `--quiet` would make sense to be handled in `opoo` directly so that all warnings are hidden, regardless of where they are coming from.

IMO `--quiet` shouldn't hide warnings. There's too much important information that's passed as warnings that I'd want to exempt from this.

File: Library/Homebrew/cask/cask.rb
Comment (Bo98) [inline Library/Homebrew/cask/cask.rb:113]: This is completely beside the point but I find it interesting there's absolutely no public API for dealing with casks.

Is dealing with casks just not allowed at all in third-party tools?

IMO, `homebrew-bundle` is an example of something that should be able to work with public API only. It is after all an example of something that was third-party made first-party, so seems silly for it to be something that third-party developers shouldn't be allowed to do.

(Sorry for the off-topic, maybe worth creating a separate discussion for this)

File: Library/Homebrew/cask/cask.rb
Comment (Bo98) [inline Library/Homebrew/cask/cask.rb:142]: A comment showing what part of the path you're going for here could be useful. Basically like what you have in `installed_version`.

File: Library/Homebrew/cask/cask.rb
Comment (Bo98) [inline Library/Homebrew/cask/cask.rb:216]: Why can `timestamped_versions` return multiple versions but this can't?

**Review by Bo98 — APPROVED** [review]

File: Library/Homebrew/cask/cask.rb
Comment (reitermarkus) [inline Library/Homebrew/cask/cask.rb:142]:
```suggestion
      # <caskroom_path>/.metadata/<version>/<timestamp>/Casks/<token>.{rb,json} -> <timestamp>
      time = installed_caskfile&.dirname&.dirname&.basename&.to_s
```

File: Library/Homebrew/cask/cask.rb
Comment (reitermarkus) [reply Library/Homebrew/cask/cask.rb:217]: `timestamped_versions` is pretty much an implementation detail at this point and probably shouldn't be public.

File: Library/Homebrew/cask/migrator.rb
Comment (reitermarkus) [reply Library/Homebrew/cask/migrator.rb:48]: Right, it was needed since `old_installed_caskfile` was absolute before. Still, it makes more sense to me to only replace it in the basename.

**Comment by reitermarkus** [comment]
Let's merge this after the next tag.

**Comment by MikeMcQuaid** [comment]
Tagged, merged, thanks @reitermarkus!

**Pull request merged**
